components/brave_shields/browser/ad_block_filters_provider_manager.cc

```diff
@@ -43,21 +43,20 @@ AdBlockFiltersProviderManager::~AdBlockFiltersProviderManager() = default;
 
 void AdBlockFiltersProviderManager::AddProvider(
     AdBlockFiltersProvider* provider) {
-  auto it =
-      std::find(filters_providers_.begin(), filters_providers_.end(), provider);
-  CHECK(it == filters_providers_.end());
-  filters_providers_.push_back(provider);
+  auto rv = filters_providers_.insert(provider);
+  DCHECK(rv.second);
   provider->AddObserver(this);
 }
 
 void AdBlockFiltersProviderManager::RemoveProvider(
     AdBlockFiltersProvider* provider) {
-  auto it =
-      std::find(filters_providers_.begin(), filters_providers_.end(), provider);
-  CHECK(it != filters_providers_.end());
-  (*it)->RemoveObserver(this);
-  filters_providers_.erase(it);
-  NotifyObservers();
+  auto it = filters_providers_.find(provider);
+  DCHECK(it != filters_providers_.end());
+  if (it != filters_providers_.end()) {
```

Comment:
from the same spec RE dcheck's, you either assume this situation is not possible (and thus put in a dcheck), or you write a condition, not both things at once